cip/1.accepted/CIP2017-04-13-Aggregations.adoc

```diff
@@ -0,0 +1,110 @@
+= CIP2017-04-13 Aggregations
+:numbered:
+:toc:
+:toc-placement: macro
+:source-highlighter: codemirror
+
+*Author:* Tobias Lindaaker <tobias.lindaaker@neotechnology.com>
+
+toc::[]
+
+== Aggregations
+
+=== Syntax
+
+[source, ebnf]
+----
+Return = 'RETURN', ['DISTINCT'], ReturnBody, Filter ;
+With = 'WITH', ['DISTINCT'], ReturnBody, Filter ;
+SingleValueReturn = 'RETURN', (Expression | ProjectedMap | Aggregation), Filter ;
+
+ReturnBody = ('*' | ReturnItem), {',', ReturnItem} ;
+ReturnItem = (Expression, ['AS', Variable])
+           | (Aggregation, ['AS', Variable])
+           | (ProjectedMap, 'AS', Variable)
+           ;
+
+Filter = [Where], [Order], [Skip], [Limit] ;
+
+Aggregation = Aggregator, 'OF', Expression ;
+Aggregator = SymbolicName | ExtensionName ;
+ExtensionName = {SymbolicName, '.'}-, SymbolicName ;
+----
+
+=== Examples
+
+[source, cypher]
+.Aggregation using `avg`
+----
+MATCH (employee:Employee)
+RETURN avg OF employee.salary
+----
+
+[source, cypher]
+.Aggregation using `collect`
+----
+MATCH (person:Person)-[:FRIEND]-(friend)
+RETURN person.email, collect OF friend {.name,.email} AS friends
+----
+
+[source, cypher]
+.Aggregation using `count`
+----
+MATCH (nodes) RETURN count OF nodes
```

Comment:
I've been wondering if we shouldn't just mandate always writing a kind-of list comprehension syntax for aggregation, to make more visible what's going on
```
collect [ n.prop | * ]
max [ n.weight | * ]
```
this would allow re-using aggregation function calling syntax with regular lists!
```
count [ n | n IN [ 1, 2, 3, 4 ] ]
```
or even shorter
```
count [ n | 1, 2, 3, 4 ]
```

I agree, we should just use `aggrF(args) ...` to pass in args to the aggregation function.

But I think it's important to have syntax that allows us to call aggregation functions for aggregations as well as over ordinary lists.